lib/src/str_util.rs

```diff
@@ -32,105 +32,189 @@ pub enum StringPatternParseError {
+    pub fn to_case_sensitive_glob(&self) -> Option<Cow<'_, str>> {
         // TODO: If we add Regex pattern, it will return None.
+        // TODO: Handle trivial case‐insensitive patterns here? It might make people
+        // expect they can use case‐insensitive patterns in contexts where they
+        // generally can’t.
         match self {
-            StringPattern::Exact(literal) => Some(glob::Pattern::escape(literal).into()),
-            StringPattern::Glob(pattern) => Some(pattern.as_str().into()),
-            StringPattern::Substring(needle) if needle.is_empty() => Some("*".into()),
-            StringPattern::Substring(needle) => {
-                Some(format!("*{}*", glob::Pattern::escape(needle)).into())
+            StringPattern::Exact(literal, CaseSensitivity::Sensitive) => {
+                Some(glob::Pattern::escape(literal).into())
+            }
+            StringPattern::Substring(needle, CaseSensitivity::Sensitive) => {
+                if needle.is_empty() {
+                    Some("*".into())
+                } else {
+                    Some(format!("*{}*", glob::Pattern::escape(needle)).into())
+                }
             }
+            StringPattern::Glob(pattern, CaseSensitivity::Sensitive) => {
+                Some(pattern.as_str().into())
+            }
+            _ => None,
         }
     }
 
     /// Returns true if this pattern matches the `haystack`.
     pub fn matches(&self, haystack: &str) -> bool {
+        // TODO: Unicode case folding is complicated and can be locale‐specific. The
+        // `glob` crate and Gitoxide only deal with ASCII case folding, so we do
+        // the same here; a more elaborate case folding system will require
+        // making sure those behave in a matching manner where relevant.
+        //
+        // Care will need to be taken regarding normalization and the choice of an
+        // appropriate case‐insensitive comparison scheme (`toNFKC_Casefold`?) to ensure
+        // that it is compatible with the standard case‐insensitivity of haystack
+        // components (like internationalized domain names in email addresses). The
+        // availability of normalization and case folding schemes in database backends
+        // will also need to be considered. A locale‐specific case folding
+        // scheme would likely not be appropriate for Jujutsu.
+        //
+        // For some discussion of this topic, see:
+        // <https://github.com/unicode-org/icu4x/issues/3151>
         match self {
-            StringPattern::Exact(literal) => haystack == literal,
-            StringPattern::Glob(pattern) => pattern.matches(haystack),
-            StringPattern::Substring(needle) => haystack.contains(needle),
+            StringPattern::Exact(literal, CaseSensitivity::Sensitive) => haystack == literal,
+            StringPattern::Exact(literal, CaseSensitivity::Insensitive) => {
+                haystack.to_ascii_lowercase() == literal.to_ascii_lowercase()
```

Comment:
nit: `eq_ignore_ascii_case()` can be used

Reply:
Thanks, I didn’t know about that!